### PR TITLE
Fix end behaviour for recent node versions

### PR DESCRIPTION
**Pull request by ZJONSSON**
This PR addresses some recent changes in the node internals for streams

### CODE REVIEW

File: streamz.js [hunk 1]
```diff
@@ -152,18 +152,30 @@ Streamz.prototype._fn = function(d) {
   this.push(d);
 };
 
-Streamz.prototype._finalize = noop;
+Streamz.prototype._finalize = function() {
+  // In node 14 we need to look at `_writableState.buffered.length` - in older versions we look at `_writableState.length`
+  if (this.endedCb && !this._concurrent && (this._writableState.buffered && !this._writableState.buffered.length || !this._writableState.length)) {
```
Comment (jbreckman) [inline streamz.js:157]: This used to only do this if incoming pipes was < 1.... but now it always does it?

File: streamz.js [hunk 2]
```diff
@@ -152,18 +152,30 @@ Streamz.prototype._fn = function(d) {
...
+    if (this._incomingPipes < 1) {
+      this.endedCb = cb || noop;
+      this._finalize();
+    } else {
+     if (cb) cb();
+    }
+  }
+
+  if (d !== undefined) {
+    this._transform(d, null, end);
```
Comment (jbreckman) [inline streamz.js:176]: oh this is kinda clever

File: streamz.js [hunk 3]
```diff
@@ -152,18 +152,30 @@ Streamz.prototype._fn = function(d) {
...
-        stream.Transform.prototype.end.apply(this, undefined, cb);
-    };
-    this._finalize();
+  
+  const end = () => {
+    if (this._incomingPipes < 1) {
+      this.endedCb = cb || noop;
+      this._finalize();
+    } else {
+     if (cb) cb();
```
Comment (jbreckman) [inline streamz.js:171]: indentation is wrong

File: streamz.js [hunk 4]
```diff
@@ -152,18 +152,30 @@ Streamz.prototype._fn = function(d) {
   this.push(d);
 };
 
-Streamz.prototype._finalize = noop;
+Streamz.prototype._finalize = function() {
+  // In node 14 we need to look at `_writableState.buffered.length` - in older versions we look at `_writableState.length`
+  if (this.endedCb && !this._concurrent && (this._writableState.buffered && !this._writableState.buffered.length || !this._writableState.length)) {
+    stream.Transform.prototype.end.apply(this, undefined, this.endedCb);
```
Comment (jbreckman) [inline streamz.js:158]: i don't see a world where this is a problem - but:

```
let endedCb = this.endedCb;
this.endedCB = noop;
stream.Transform.prototype.end.apply(this, undefined, endedCb);
```
would be technically more correct.  It's possible (but unlikely) the `end` method somehow modifies `endedCb`

File: streamz.js (hunk 1 above)
Comment (ZJONSSON) [reply streamz.js:157]: `endedCb` is only set if incomingPipes < 1  https://github.com/ZJONSSON/streamz/pull/25/commits/570d6f1e8f288e461b99016505cbe266dadb7376#diff-4823b7de66d03ac4621b2d4cd22ede1e1fd8df119a3fc293dc0e2f0421b024e0R167-R169

File: streamz.js (hunk 4 above)
Comment (ZJONSSON) [reply streamz.js:158]: since we are passing the actual `this.endedCb` but not a reference I agree this shouldn't be a problem, but good point though!

**Comment by ZJONSSON** [comment]
Thanks @jbreckman

File: streamz.js (hunk 1 above)
Comment (jbreckman) [reply streamz.js:157]: oh interesting.  i didn't catch that.  so now it checks for incoming pipes < 1 after the last transform is finished, which is presumably better.

**Review by jbreckman — COMMENTED** [review]
lgtm

**Pull request merged**